### PR TITLE
optimize ThreadlessExecutor

### PR DESCRIPTION
**Pull request by icodening**
## What is the purpose of the change
1. optimize ThreadlessExecutor
2. tri response works on user threads

The original `ThreadlessExecutor` uses the `LinkedBlockingQueue.take` method to wait response, which will reduce the performance of the benchmark.

## Checklist
- [x] Make sure there is a [GitHub_issue](https://github.com/apache/dubbo/issues) field for the change (usually before you start working on it). Trivial changes like typos do not require a GitHub issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue.
- [ ] Each commit in the pull request should have a meaningful subject line and body.
- [ ] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [ ] Check if is necessary to patch to Dubbo 3 if you are work on Dubbo 2.7
- [ ] Write necessary unit-test to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add sample in [dubbo samples](https://github.com/apache/dubbo-samples) project.
- [ ] Add some description to [dubbo-website](https://github.com/apache/dubbo-website) project if you are requesting to add a feature.
- [ ] GitHub Actions works fine on your own branch.
- [ ] If this contribution is large, please follow the [Software Donation Guide](https://github.com/apache/dubbo/wiki/Software-donation-guide).

### CODE REVIEW

File: dubbo-common/src/main/java/org/apache/dubbo/common/threadpool/ThreadlessExecutor.java [hunk 1]
```diff
@@ -56,60 +58,37 @@ public void setWaitingFuture(CompletableFuture<?> waitingFuture) {
...
 
-    private boolean isFinished() {
-        return finished;
-    }
-
-    private void setFinished(boolean finished) {
-        this.finished = finished;
-    }
-
     public boolean isWaiting() {
```
Comment (icodening) [inline dubbo-common/src/main/java/org/apache/dubbo/common/threadpool/ThreadlessExecutor.java:61]: no need here, to be considered

**Review by AlbumenJ — APPROVED** [review]

File: dubbo-rpc/dubbo-rpc-triple/src/main/java/org/apache/dubbo/rpc/protocol/tri/TripleInvoker.java [hunk 2]
```diff
@@ -127,12 +128,14 @@ protected Result doInvoke(final Invocation invocation) {
...
         ClientCall call = new TripleClientCall(connectionClient, streamExecutor,
             getUrl().getOrDefaultFrameworkModel(), writeQueue);
-
+        Executor callbackExecutor = getCallbackExecutor(getUrl(), invocation);
         AsyncRpcResult result;
         try {
             switch (methodDescriptor.getRpcType()) {
                 case UNARY:
-                    result = invokeUnary(methodDescriptor, invocation, call);
+                    call = new TripleClientCall(connectionClient, callbackExecutor,
```
Comment (EarthChen) [inline dubbo-rpc/dubbo-rpc-triple/src/main/java/org/apache/dubbo/rpc/protocol/tri/TripleInvoker.java:136]: The call should not be new twice

**Review by EarthChen — CHANGES_REQUESTED** [review]

**Review by EarthChen — APPROVED** [review]
LGTM

File: dubbo-common/src/main/java/org/apache/dubbo/common/threadpool/ThreadlessExecutor.java [hunk 3]
```diff
@@ -38,78 +39,39 @@
...
-    }
-
-    public boolean isWaiting() {
-        return waiting;
-    }
-
-    private void setWaiting(boolean waiting) {
-        this.waiting = waiting;
-    }
+    private volatile Object waiter;
```
Comment (icodening) [inline dubbo-common/src/main/java/org/apache/dubbo/common/threadpool/ThreadlessExecutor.java:46]: Only need to be visible to other threads

**Comment by vio-lin** [comment]
timeout 超时回调加了一个 isShutdown的判断。  
我们线上也出现了 线程池被关掉的情况。 问下当时是有什么逻辑 触发了这个shutdonw 吗。
```
  ExecutorService executor = future.getExecutor();
            if (executor != null && !executor.isShutdown()) {
                executor.execute(() -> notifyTimeout(future));
            } else {
                notifyTimeout(future);
            }
```

**Pull request merged**